_posts/2015-07-21-build.md

```diff
@@ -0,0 +1,55 @@
+
+### mach bootstrap
+
+The steps that mach will handle before issuing a normal `cargo build` command are:
+* Downloading the correct versions of the cargo and rustc tools. Servo uses most of the unstable features in Rust, which change pretty frequently. We also test the edges of feature compatibility and so are the first ones to notice many changes that did not at first seem as if they would break anyone. Further, we build a custom version of the tools that additionally supports cross-compilation targeting Android (and ARM in the near future). A random local install of the Rust toolchain is pretty unlikely to work with Servo.
+
+* Updating git submodules. Some of Servo's dependencies cannot be downloaded as Cargo dependencies because they need to be directly referenced in the build process, and Cargo adds a hash that makes it difficult to locate those files. For such code, we add them as submodules.
+
+### mach build & run
+
+The build itself also verifies that the user has explicitly requested either a `dev` or `release` build --- the Servo dev build is debuggable but quite slow, and it's not clear which build should be the default.
+
+Additionally, there's the question of _which_ `cargo build` to run. Servo has three different "toplevel" Cargo.toml files.
+* `components/servo/Cargo.toml` is used to build an executable binary named `servo` and is used on Linux and OSX. There are also horrible linker hacks in place that will cause an Android-targeted build to instead produce a file named `servo` that is actually an APK file that can be loaded onto Android devices.
+* `ports/gonk/Cargo.toml` produces a binary that can run on the Firefox OS Boot2Gecko mobile platform.
+* `ports/cef/Cargo.toml` produces a shared library that can be loaded within the Chromium Embedding Framework to provide a hostable web rendering engine.
+
+The presence of these three different toplevel binaries and the curious directory structure means that mach also provides a `run` command that will execute the correct binary with any provided arguments.
+
+### mach test
+
+Servo has several testing tools that can be executed via mach.
+
+* `mach tidy` will verify that there are no trivial syntactic errors in source files. It checks for valid license headers in each file, no tab characters, no trailing whitespaces, etc.
+
+* `mach test-ref` will run the Servo-specific reference tests. These tests render a pair of web pages that implement the same final layout using different CSS features to images. If the images are not pixel-identically, the test fails.
+
+* `mach test-wpt` runs the cross-browser W3C Web Platform Tests, which primarily test DOM features.
+
+* `mach test-css` runs the cross-browser CSS WG reference tests, which are a version of the reference tests that are intended to work across many browsers.
+
+* `mach test-unit` runs the Rust unit tests embedded in Servo crates. We do not have many of these, except for basic tests of per-crate functionality.
```

Comment:
Might be worth mentioning that we don't have many of these because wpt and css cover most bases.